docs/protocol_upgrade.md

```diff
@@ -0,0 +1,150 @@
+            #[cfg(feature = "protocol_feature_evm")]
+            ProtocolFeature::EVM => 100,
+        }
+    }
+}  
+```
+- When we need to decide whether to use the new feature based on the protocol version of the current network,
+  we can simply compare it to the protocol version of the feature. For example,
+```rust
+#[cfg(feature = “protocol_version_evm”)]
+if current_protocol_version >= ProtocolFeature::EVM.protocol_version() {
+   do_evm_change();
+} else {
+   do_old_stuff();
+}
+#[cfg(not(feature = “protocol_version_evm”))] {
+   do_old_stuff();
+}
+```
+To make this simpler, we write a macro `checked_feature` that conveniently abstracts this logic:
+```rust
+#[macro_export]
+macro_rules! checked_feature {
+    ($feature_name:tt, $feature:ident, $current_protocol_version:expr) => {{
+        #[cfg(feature = $feature_name)]
+        let is_feature_enabled = $crate::version::ProtocolFeature::$feature.protocol_version()
+            <= $current_protocol_version;
+        #[cfg(not(feature = $feature_name))]
+        let is_feature_enabled = {
+            // Workaround unused variable warning
+            let _ = $current_protocol_version;
+
+            false
+        };
+        is_feature_enabled
+    }};
+
+    ($feature_name:tt, $feature:ident, $current_protocol_version:expr, $feature_block:block) => {{
+        checked_feature!($feature_name, $feature, $current_protocol_version, $feature_block, {})
+    }};
+
+    ($feature_name:tt, $feature:ident, $current_protocol_version:expr, $feature_block:block, $non_feature_block:block) => {{
+        #[cfg(feature = $feature_name)]
+        {
+            if checked_feature!($feature_name, $feature, $current_protocol_version) {
+                $feature_block
+            } else {
+                $non_feature_block
+            }
+        }
+        // Workaround unused variable warning
+        #[cfg(not(feature = $feature_name))]
+        {
+            let _ = $current_protocol_version;
+            $non_feature_block
+        }
+    }};
+}
+```
+With this macro, the previous example can be simplified into
+```rust
+checked_feature!(“protocol_feature_evm”, EVM, {do_evm_stuff();}, {do_old_stuff();});
+```
+
+### Testing
+
+Nightly protocol features allow us to enable the most bleeding-edge code in some testing environment. We can choose to
+enable all nightly protocol features by
+```
+cargo build -p neard --release --features nightly_protocol_features
+```
+or enable some specific protocol feature by
+```
+cargo build -p neard --release --features nightly_protocol,<protocol_feature>
+```
+
+In practice, we have all nightly protocol features enabled for Nayduck tests and on betanet, which is updated daily.
+
+### Feature Stabilization
+
+When we are ready to make a new release, we examine which features should be stabilized and for those that should be, we
+remove the feature gating everywhere. One difference is that we now use this rule of the `checked_feature` macro:
```

Comment:
This is a bit confusing. We still have protocol version checks, because we need to maintain behavior for old blocks. It's just that we remove cargo featrures. 

I think it maybe makes sense to call the two kinds of feature gating at the start of the document:

* for stable features, we want to check if the features is enabled for the block's protocol version
* for unstable features, we shoudl check if we are runing nightly binary as well. 

Hm, now that I've spelled this out, it seems like it would be useful to discuss where protocol version comes from. THat is, that we associated each block with a protocol version at which it was produced.